### ISSUE TITLE
Freeing of bin can cause a double free if CUDA initialized with program_init_binary()

### ISSUE THREAD

**Issue by mrutt92** [bespoke-silicon-group/bsg_f1#337]
Suppose CUDA is initialized with:
```c
hb_mc_device_progam_init_binary(hb_mc_device_t *device, 
                                     const char *bin_name,
                                     const unsigned char* bin_data, 
                                     size_t bin_size, 
                                     const char* alloc_name, 
                                     hb_mc_allocator_id_t id)
```
This line of code will cause an invalid free: the consequences of this are dire.

https://github.com/bespoke-silicon-group/bsg_f1/blob/39ae1eb895d0bdc3010dca6d8821546d8e5ff244/cl_manycore/libraries/bsg_manycore_cuda.cpp#L1236

**Comment by drichmond** [comment]
Was this ever addressed?

**Comment by bornaehsani** [comment]
> 
> 
> Was this ever addressed?

Oh god I totally forgot about this. I submitted a PR for the fix, plus a regression test for it.

**Comment by bornaehsani** [comment]
Addressed in PR #419.
